File: ai-solutions/04/sonnet-4.5/python/main.py
import os

def solve():
    # Read input file
    input_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '..', '..', 'inputs', '04.txt')
    input_path = os.path.normpath(input_path)
    with open(input_path, 'r') as f:
        input_text = f.read().strip()
    
    # Parse the grid
    grid = [list(line) for line in input_text.split('\n')]
    rows = len(grid)
    cols = len(grid[0])
    
    # Helper function to count neighbors of a position
    def count_neighbors(grid, r, c):
        directions = [
            (-1, -1), (-1, 0), (-1, 1),
            (0, -1),           (0, 1),
            (1, -1),  (1, 0),  (1, 1)
        ]
        
        count = 0
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            if 0 <= nr < len(grid) and 0 <= nc < len(grid[0]):
                if grid[nr][nc] == '@':
                    count += 1
        return count
    
    # Part 1: Count initially accessible rolls
    def part1(grid):
        accessible = 0
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == '@':
                    if count_neighbors(grid, r, c) < 4:
                        accessible += 1
        return accessible
    
    # Part 2: Iteratively remove accessible rolls
    def part2(grid):
        # Create a mutable copy of the grid
        mutable_grid = [row[:] for row in grid]
        total_removed = 0
        
        while True:
            # Find all accessible rolls
            to_remove = []
            for r in range(len(mutable_grid)):
                for c in range(len(mutable_grid[0])):
                    if mutable_grid[r][c] == '@':
                        if count_neighbors(mutable_grid, r, c) < 4:
                            to_remove.append((r, c))
            
            # If no rolls can be removed, stop
            if not to_remove:
                break
            
            # Remove all accessible rolls
            for r, c in to_remove:
                mutable_grid[r][c] = '.'
            
            total_removed += len(to_remove)
        
        return total_removed
    
    # Solve both parts
    answer1 = part1(grid)
    answer2 = part2(grid)
    
    print(answer1)
    print(answer2)

if __name__ == '__main__':
    solve()

File: ai-solutions/04/sonnet-4.5/rust/src/main.rs
use std::fs;
use std::path::PathBuf;

fn count_neighbors(grid: &Vec<Vec<char>>, r: usize, c: usize) -> usize {
    let directions = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1),           (0, 1),
        (1, -1),  (1, 0),  (1, 1),
    ];
    
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;
    let mut count = 0;
    
    for (dr, dc) in directions.iter() {
        let nr = r as i32 + dr;
        let nc = c as i32 + dc;
        
        if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
            if grid[nr as usize][nc as usize] == '@' {
                count += 1;
            }
        }
    }
    
    count
}

fn part1(grid: &Vec<Vec<char>>) -> usize {
    let mut accessible = 0;
    
    for r in 0..grid.len() {
        for c in 0..grid[0].len() {
            if grid[r][c] == '@' {
                if count_neighbors(grid, r, c) < 4 {
                    accessible += 1;
                }
            }
        }
    }
    
    accessible
}

fn part2(grid: &Vec<Vec<char>>) -> usize {
    // Create a mutable copy of the grid
    let mut mutable_grid: Vec<Vec<char>> = grid.iter()
        .map(|row| row.clone())
        .collect();
    
    let mut total_removed = 0;
    
    loop {
        // Find all accessible rolls
        let mut to_remove = Vec::new();
        
        for r in 0..mutable_grid.len() {
            for c in 0..mutable_grid[0].len() {
                if mutable_grid[r][c] == '@' {
                    if count_neighbors(&mutable_grid, r, c) < 4 {
                        to_remove.push((r, c));
                    }
                }
            }
        }
        
        // If no rolls can be removed, stop
        if to_remove.is_empty() {
            break;
        }
        
        // Remove all accessible rolls
        for (r, c) in &to_remove {
            mutable_grid[*r][*c] = '.';
        }
        
        total_removed += to_remove.len();
    }
    
    total_removed
}

fn main() {
    // Read input file
    let mut input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    input_path.pop(); // rust
    input_path.pop(); // sonnet-4.5
    input_path.pop(); // 04
    input_path.pop(); // ai-solutions
    input_path.push("inputs");
    input_path.push("04.txt");
    
    let input = fs::read_to_string(input_path)
        .expect("Failed to read input file");
    
    // Parse the grid
    let grid: Vec<Vec<char>> = input
        .trim()
        .lines()
        .map(|line| line.chars().collect())
        .collect();
    
    // Solve both parts
    let answer1 = part1(&grid);
    let answer2 = part2(&grid);
    
    println!("{}", answer1);
    println!("{}", answer2);
}

